HoRNDIS.h

```diff
@@ -226,104 +239,128 @@ struct rndis_set_c {
 /***** Actual class definitions *****/
 
 typedef struct {
-	bool inuse;
 	IOBufferMemoryDescriptor *mdp;
-	void *buf;
-	IOUSBCompletion comp;
+	IOUSBHostCompletion comp;
 } pipebuf_t;
 
 class HoRNDIS : public IOEthernetController {
 	OSDeclareDefaultStructors(HoRNDIS);	// Constructor & Destructor stuff
 
 private:
-	bool fTerminate; // being terminated now (i.e., device being unplugged)
+	/*!
+	 * This class protects method calls against re-entry, when the IOCommand
+	 * gate is being released due to synchronous IO, or some other reason.
+	 * Use case:
+	 * Unlike start/stop calls, that are triggered by a single IO provider,
+	 * the enable/disable calls can be triggered by multiple interface clients,
+	 * as well as user's actions, e.g. "ifconfig en6 up" - potentially multiple
+	 * "ifconfig" processes running in parallel.
+	 * Even though the calls to this class are protected by the IOCommandGate,
+	 * synchronous USB transfers release the gate (by using 
+	 * IOCommandGate::commandSleep), allowing another enable/disable call
+	 * to "sneak in". We use the "ReentryLocker" to delay additional
```

Comment:
This explains some of this (see comment above; citing this comment in the .cpp implementation might be good too) but I am still wondering why deadlock cannot happen...

Reply:
See above comment any my email